File: apps/syn-cli/src/syn_cli/commands/sessions.py
"""Session management commands — list and show sessions."""

from __future__ import annotations

import typer
from rich.panel import Panel
from rich.table import Table

from syn_cli._output import (
    console,
    format_cost,
    format_timestamp,
    format_tokens,
    status_style,
)
from syn_cli.commands._api_helpers import api_get, api_get_list, build_params
from syn_cli.commands._session_models import SessionDetailResponse, SessionSummaryResponse

app = typer.Typer(
    name="sessions",
    help="View and inspect agent sessions",
    no_args_is_help=True,
)


@app.command("list")
def list_sessions(
    workflow_id: str | None = typer.Option(None, "--workflow", "-w", help="Filter by workflow ID"),
    status: str | None = typer.Option(None, "--status", "-s", help="Filter by status"),
    limit: int = typer.Option(50, "--limit", "-n", help="Max results", min=1, max=200),
) -> None:
    """List agent sessions."""
    params = build_params(workflow_id=workflow_id, status=status, limit=limit)
    items = [SessionSummaryResponse(**s) for s in api_get_list("/sessions", params=params)]

    if not items:
        console.print("[dim]No sessions found.[/dim]")
        return

    table = Table(title="Sessions")
    table.add_column("ID", style="dim")
    table.add_column("Status")
    table.add_column("Provider")
    table.add_column("Tokens", justify="right")
    table.add_column("Cost", justify="right")
    table.add_column("Started", justify="right")

    for s in items:
        style = status_style(s.status)
        table.add_row(
            s.id[:12] + "..." if len(s.id) > 12 else s.id,
            f"[{style}]{s.status}[/{style}]" if style else s.status,
            s.agent_provider or "-",
            format_tokens(s.total_tokens),
            format_cost(s.total_cost_usd),
            format_timestamp(s.started_at),
        )
    console.print(table)


def _build_session_panel(s: SessionDetailResponse) -> str:
    """Build panel text for session detail view."""
    style = status_style(s.status)
    status_display = f"[{style}]{s.status}[/{style}]" if style else s.status
    text = (
        f"[bold]Session:[/bold] {s.id}\n"
        f"[bold]Status:[/bold] {status_display}\n"
        f"[bold]Provider:[/bold] {s.agent_provider or '-'} / {s.agent_model or '-'}\n"
        f"[bold]Workflow:[/bold] {s.workflow_name or s.workflow_id or '-'}\n"
        f"[bold]Tokens:[/bold] {format_tokens(s.total_tokens)} "
        f"(in: {format_tokens(s.input_tokens)}, out: {format_tokens(s.output_tokens)})\n"
        f"[bold]Cost:[/bold] {format_cost(s.total_cost_usd)}\n"
        f"[bold]Started:[/bold] {format_timestamp(s.started_at)}"
    )
    if s.error_message:
        text += f"\n[bold red]Error:[/bold red] {s.error_message}"
    return text


@app.command("show")
def show_session(
    session_id: str = typer.Argument(..., help="Session ID"),
) -> None:
    """Show detailed information for a session."""
    s = SessionDetailResponse(**api_get(f"/sessions/{session_id}"))

    console.print(
        Panel(
            _build_session_panel(s),
            title="[cyan]Session Detail[/cyan]",
            border_style="cyan",
        )
    )

    if s.operations:
        table = Table(title="Operations")
        table.add_column("Type")
        table.add_column("Tool", style="cyan")
        table.add_column("Duration", justify="right")
        table.add_column("Status")

        for op in s.operations:
            dur = f"{op.duration_seconds:.1f}s" if op.duration_seconds else "-"
            ok = "[green]ok[/green]" if op.success else "[red]fail[/red]"
            table.add_row(op.operation_type, op.tool_name or "-", dur, ok)
        console.print(table)

File: apps/syn-cli-node/src/commands/sessions.ts
/**
 * Session management commands - list, show.
 *
 * Reads server-produced *_display fields for human-readable values
 * (tokens, cost, duration, model). Locale-dependent timestamps are
 * formatted client-side from the API's ISO 8601 UTC strings.
 *
 * See: docs/adrs/ADR-064-observability-monitor-ui.md
 */

import { CommandGroup, type CommandDef, type ParsedArgs } from "../framework/command.js";
import { CLIError } from "../framework/errors.js";
import { api, unwrap } from "../client/typed.js";
import type { components } from "../generated/api-types.js";
import { print, printError, printDim } from "../output/console.js";
import { style, BOLD, CYAN, DIM } from "../output/ansi.js";
import { formatStatus, formatTimestamp } from "../output/format.js";
import { Table } from "../output/table.js";

type SessionSummary = components["schemas"]["SessionSummaryResponse"];
type SessionDetail = components["schemas"]["SessionResponse"];

const listCommand: CommandDef = {
  name: "list",
  description: "List agent sessions",
  options: {
    workflow: { type: "string", short: "w", description: "Filter by workflow ID" },
    status: { type: "string", short: "s", description: "Filter by status" },
    limit: { type: "string", short: "n", description: "Max results", default: "50" },
  },
  handler: async (parsed: ParsedArgs) => {
    const workflow = parsed.values["workflow"] as string | undefined;
    const status = parsed.values["status"] as string | undefined;
    const limitStr = (parsed.values["limit"] as string | undefined) ?? "50";

    const data = unwrap(await api.GET("/sessions", {
      params: {
        query: {
          workflow_id: workflow ?? null,
          status: status ?? null,
          limit: parseInt(limitStr, 10),
        },
      },
    }), "Failed to list sessions");

    const items: SessionSummary[] = data.sessions ?? [];
    if (items.length === 0) { printDim("No sessions found."); return; }

    const table = new Table({ title: "Sessions" });
    table.addColumn("Session ID", { style: CYAN });
    table.addColumn("Status");
    table.addColumn("Model");
    table.addColumn("Started");
    table.addColumn("Tokens", { align: "right" });
    table.addColumn("Cost", { align: "right" });

    for (const s of items) {
      table.addRow(
        s.id.slice(0, 8) + "\u2026",
        formatStatus(s.status),
        s.agent_model_display ?? s.agent_provider ?? "\u2014",
        formatTimestamp(s.started_at),
        s.total_tokens_display,
        s.total_cost_display,
      );
    }
    table.print();
  },
};

const showCommand: CommandDef = {
  name: "show",
  description: "Show detailed session information",
  args: [{ name: "session-id", description: "Session ID", required: true }],
  handler: async (parsed: ParsedArgs) => {
    const id = parsed.positionals[0];
    if (!id) { printError("Missing session-id"); throw new CLIError("Missing argument", 1); }

    const d: SessionDetail = unwrap(await api.GET("/sessions/{session_id}", {
      params: { path: { session_id: id } },
    }), "Failed to get session");

    print(`${style("Session:", BOLD)} ${d.id}`);
    print(`  Workflow:    ${d.workflow_name ?? d.workflow_id ?? "\u2014"}`);
    print(`  Status:      ${formatStatus(d.status)}`);
    print(`  Provider:    ${d.agent_provider ?? "\u2014"}`);
    print(`  Model:       ${d.agent_model_display ?? d.agent_model ?? "\u2014"}`);
    print(`  Started:     ${formatTimestamp(d.started_at)}`);
    if (d.completed_at) print(`  Completed:   ${formatTimestamp(d.completed_at)}`);
    if (d.duration_seconds != null) print(`  Duration:    ${d.duration_display}`);
    print(`  Tokens:      ${d.total_tokens_display} (in: ${d.input_tokens_display}, out: ${d.output_tokens_display})`);
    if (d.cache_creation_tokens) print(`  Cache Write: ${d.cache_creation_tokens_display}`);
    if (d.cache_read_tokens) print(`  Cache Read:  ${d.cache_read_tokens_display}`);
    print(`  Cost:        ${d.total_cost_display}`);
    if (d.error_message) print(`  Error:       ${d.error_message}`);

    const operations = d.operations ?? [];
    if (operations.length > 0) {
      print("");
      const table = new Table({ title: "Operations" });
      table.addColumn("#", { align: "right", style: DIM });
      table.addColumn("Type", { style: CYAN });
      table.addColumn("Tool");
      table.addColumn("Success");

      for (let i = 0; i < operations.length; i++) {
        const op = operations[i]!;
        table.addRow(
          String(i + 1),
          op.operation_type,
          op.tool_name ?? "\u2014",
          op.success ? "yes" : "no",
        );
      }
      table.print();
    }
  },
};

export const sessionsGroup = new CommandGroup("sessions", "List and inspect agent sessions");
sessionsGroup.command(listCommand).command(showCommand);
